src/Transform/ONNX/ReplaceNoValueOperands.cpp

```diff
@@ -0,0 +1,141 @@
+#include "src/Dialect/ONNX/ONNXOps.hpp"
+#include "src/Dialect/ONNX/ONNXOps/OpHelper.hpp"
+#include "src/Pass/Passes.hpp"
+
+using namespace mlir;
+
+namespace onnx_mlir {
+namespace {
+struct ReplaceNoValuePass : public mlir::PassWrapper<ReplaceNoValuePass,
+                                OperationPass<mlir::ModuleOp>> {
+  MLIR_DEFINE_EXPLICIT_INTERNAL_INLINE_TYPE_ID(ReplaceNoValuePass)
+
+  ReplaceNoValuePass() = default;
+  ReplaceNoValuePass(const ReplaceNoValuePass &pass)
+      : PassWrapper<ReplaceNoValuePass, OperationPass<ModuleOp>>() {}
+
+  StringRef getArgument() const override { return "onnx-replace-novalue"; }
+
+  StringRef getDescription() const override {
+    return "Replaces to NoValue op with an constant op of the standard shape "
+           "and value for the op.";
+  }
+  void runOnOperation() final;
+};
+
+bool isNotNoValue(Value &value) { return !value.getType().isa<NoneType>(); }
+
+Value createONNXConstFromFloatValue(PatternRewriter &rewriter,
+    const Location &loc, ArrayRef<int64_t> shape, float floatValue) {
+  DenseElementsAttr newConstantAttr = DenseElementsAttr::get(
+      RankedTensorType::get(shape, rewriter.getF32Type()), {floatValue});
+  auto constOp =
+      createONNXConstantOpWithDenseAttr(rewriter, loc, newConstantAttr);
+  return constOp;
+}
+
+class PadReplaceNoValue : public OpRewritePattern<ONNXPadOp> {
+public:
+  using OpRewritePattern<ONNXPadOp>::OpRewritePattern;
+  LogicalResult matchAndRewrite(
+      ONNXPadOp op, PatternRewriter &rewriter) const override {
+    auto constValue = op.constant_value();
+    if (isNotNoValue(constValue)) {
+      return success();
+    }
+
+    constValue =
+        createONNXConstFromFloatValue(rewriter, op.getLoc(), {1}, 0.0F);
+    op.setOperand(2, constValue);
+    return success();
+  }
+};
+
+class GemmReplaceNoValue : public OpRewritePattern<ONNXGemmOp> {
+public:
+  using OpRewritePattern<ONNXGemmOp>::OpRewritePattern;
+  LogicalResult matchAndRewrite(
+      ONNXGemmOp op, PatternRewriter &rewriter) const override {
+    auto matrixC = op.C();
+    if (isNotNoValue(matrixC)) {
+      return success();
+    }
+
+    matrixC = createONNXConstFromFloatValue(rewriter, op.getLoc(), {1}, 0.0F);
+    op.setOperand(2, matrixC);
+    return success();
+  }
+};
+
+class Conv2DReplaceNoValue : public OpRewritePattern<ONNXConvOp> {
+public:
+  using OpRewritePattern<ONNXConvOp>::OpRewritePattern;
+  LogicalResult matchAndRewrite(
+      ONNXConvOp op, PatternRewriter &rewriter) const override {
+
+    auto bias = op.B();
+    if (isNotNoValue(bias)) {
+      return success();
+    }
+
+    auto weight = op.W();
+    auto weightType = weight.getType().cast<ShapedType>();
+    auto weightShape = weightType.getShape();
+
+    bias = createONNXConstFromFloatValue(
+        rewriter, op.getLoc(), {weightShape[0]}, 0.0F);
+
+    op->setOperand(2, bias);
+    return success();
+  }
+};
+
+void ReplaceNoValuePass::runOnOperation() {
```

Comment:
Wasn't there some discussions at one point about replacing all uses of a NoValue? What happens here if two different operators are using the same NoValue? it will only replace it for the first one, if my understanding is correct right?

Reply:
The past will replace all uses of a NoValue with a const by going through each op that matches the pattern. The folding then will delete the NoValue if there are no users left

Reply:
Ok my bad then, that wasn't really clear at a first glance. Can we have a test for this case?